### PR TITLE
Added integer range phpdoc support

### PR DESCRIPTION
**Pull request by clxmstaab**
as suggested in https://github.com/phpstan/phpstan/issues/3383#issuecomment-888080019 I added a failling unit test and started to look arround in `TypeNodeResolver`.

### CODE REVIEW

**Comment by clxmstaab** [comment]
I have not really a idea where to start... could someone give me a hint?

the test is currently failling with 

```
There was 1 failure:

1) PHPStan\Analyser\NodeScopeResolverTest::testFileAsserts with data set "C:\dvl\GitHub\phpstan-src\tests\PHPStan\Analyser/data/integer-range-types.php:167" ('type', 'C:\dvl\GitHub\phpstan-src\tes...es.php', PHPStan\Type\Constant\ConstantStringType Object (...), PHPStan\Type\ErrorType Object (...), 167)
Expected type int<0, 100>, got type *ERROR* in C:\dvl\GitHub\phpstan-src\tests\PHPStan\Analyser/data/integer-range-types.php on line 167.
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'int<0, 100>'
+'*ERROR*'

C:\dvl\GitHub\phpstan-src\src\Testing\TypeInferenceTestCase.php:124
C:\dvl\GitHub\phpstan-src\tests\PHPStan\Analyser\NodeScopeResolverTest.php:466

```

**Comment by ondrejmirtes** [comment]
As I said it needs to be implemented in TypeNodeResolver, specifically in `resolveGenericTypeNode`. It needs to resolve to `IntegerRangeType`.

**Comment by clxmstaab** [comment]
> As I said it needs to be implemented in TypeNodeResolver

sorry, I mixed up `TypeNodeResolver` and `NodeScopeResolver` :-/

> specifically in `resolveGenericTypeNode`. It needs to resolve to `IntegerRangeType`.

thx for the hint, this was really usefull

I was able to implement the minmal test-cases which came to my mind. I am pretty sure you will not like the if-cases arround the `GenericClassStringType` class-string comparison, but maybe you can give me another hint, how this could be improved

File: src/PhpDoc/TypeNodeResolver.php [hunk 1]
```diff
@@ -474,6 +475,34 @@ private function resolveGenericTypeNode(GenericTypeNode $typeNode, NameScope $na
...
+		} elseif ($mainTypeName === 'int') {
+			if (count($genericTypes) === 2) { // int<min, max>, int<1, 3>
+
+				$minType = $genericTypes[0];
+				$maxType = $genericTypes[1];
+
+				if ($minType instanceof ConstantIntegerType) {
+					$min = $minType->getValue();
+				} elseif ($minType instanceof ObjectType) {
+					if ($minType->getClassName() === "PHPStan\Generics\GenericClassStringType\min") {
```
Comment (ondrejmirtes) [inline src/PhpDoc/TypeNodeResolver.php]: This doesn't make sense, it's not going to work in a different namespace. Try looking for IdentifierTypeNode in `$typeNode->genericTypes` instead.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/PhpDoc/TypeNodeResolver.php (hunk 1 above)
Comment (clxmstaab) [reply src/PhpDoc/TypeNodeResolver.php]: lol, after reading your comment, I realized why the class-name contained this PHPStan prefixed string, which I initially thought of is a magic-value ;-)

changed the impl to use `IdentifierTypeNode` and added a few more testcases.

thanks again for your input

File: src/Dependency/ExportedNode/ExportedClassConstantNode.php [hunk 2]
```diff
@@ -83,7 +83,7 @@ public static function decode(array $data): ExportedNode
 			$data['public'],
 			$data['private'],
 			$data['final'],
-			$data['phpDoc'],
+			$data['phpDoc'] !== null ? ExportedPhpDocNode::decode($data['phpDoc']['data']) : null
```
Comment (staabm) [inline src/Dependency/ExportedNode/ExportedClassConstantNode.php:86]: Fyi, This change was not added by me

**Comment by ondrejmirtes** [comment]
Just rebased and force-pushed this. Thank you!

**Review by ondrejmirtes — APPROVED** [review]

**Pull request merged**